### PR TITLE
Deprecate `LightningModule.datamodule` reference in favor of the trainer one (#6929)

### LINKED ISSUE #6929: Missing LightningModule datamodule reference
## 🐛 Bug

This docs snippet does not work:

https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#datamodule

### To Reproduce

```python
def test_bug(tmpdir):
    class TestModel(BoringModel):
        def configure_optimizers(self):
            # works
            len(self.trainer.datamodule.train_dataloader())
            # does not
            len(self.datamodule.train_dataloader())
            
            return super().configure_optimizers()

    dm = BoringDataModule()
    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True)
    trainer.fit(TestModel(), datamodule=dm)
```

```console
line 128, in configure_optimizers
    num_training_samples = len(self.datamodule.train_dataloader())
AttributeError: 'NoneType' object has no attribute 'train_dataloader'
```

### Expected behavior

`self.datamodule` works

## Alternatives

Update docs

### Environment

master

### Additional context

Reported by Brian Staber in Slack

### PR DESCRIPTION
**Pull request by mesejo**
## What does this PR do?

This PR attempts to deprecate datamodule and update the documentation accordingly as decided in the GitHub issue

Fixes #6929

### CODE REVIEW

File: docs/source/common/lightning_module.rst [hunk 1]
```diff
@@ -985,10 +985,13 @@ datamodule
 ~~~~~~~~~~
 Set or access your datamodule.
 
+.. warning::
+    ``datamodule`` is deprecated from v1.3 and will be removed in v1.5. Access it through the trainer.
+
```
Comment (carmocca) [inline docs/source/common/lightning_module.rst:990]: This warning is not needed here as you updated the snippet below

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -163,6 +164,7 @@ def example_input_array(self, example: Any) -> None:
         self._example_input_array = example
 
     @property
+    @deprecated(deprecated_in="1.3.0", remove_in="1.5.0", target=None)
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:167]: I think it's better to use `rank_zero_warn` here as there is not target.

Also for the setter

cc: @Borda

**Review by carmocca — CHANGES_REQUESTED** [review]
Can you also add a test to https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/deprecated_api/test_remove_1-5.py ?

**Comment by awaelchli** [comment]
added the requested test

**Review by edgarriba — APPROVED** [review]

File: docs/source/common/lightning_module.rst (hunk 1 above)
Comment (carmocca) [reply docs/source/common/lightning_module.rst:996]:
```suggestion
```

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -162,6 +162,10 @@ def example_input_array(self, example: Any) -> None:
 
     @property
     def datamodule(self) -> Any:
+        rank_zero_deprecation(
+            "The `LightningModule.datamodule` getter and setter are deprecated and will be removed in v1.5."
+            " Access the datamodule through the Trainer instead."
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:167]:
```suggestion
            " Access the datamodule through using `self.trainer.datamodule` instead."
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -162,6 +162,10 @@ def example_input_array(self, example: Any) -> None:
 
     @property
     def datamodule(self) -> Any:
+        rank_zero_deprecation(
+            "The `LightningModule.datamodule` getter and setter are deprecated and will be removed in v1.5."
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:166]:
```suggestion
            "The `LightningModule.datamodule` property is deprecated in v1.3 and will be removed in v1.5."
```

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Pull request merged**